File: Hard/1326. Minimum Number of Taps to Open to Water a Garden/1326. Minimum Number of Taps to Open to Water a Garden.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        coverage = [0] * (n + 1)

        # Update coverage list based on tap ranges
        for i in range(len(ranges)):
            if ranges[i] == 0:
                continue
            left = max(0, i - ranges[i])
            coverage[left] = max(coverage[left], i + ranges[i])

        end = 0
        farthest_can_reach = 0
        tap_count = 0

        # Traverse the garden and calculate required taps
        for i in range(n + 1):
            if i > end:
                if farthest_can_reach <= end:
                    return -1
                end = farthest_can_reach
                tap_count += 1
            farthest_can_reach = max(farthest_can_reach, coverage[i])

        # Return the total taps required, considering the endpoint
        return tap_count + (1 if end < n else 0)

File: Hard/1326. Minimum Number of Taps to Open to Water a Garden/1326. Minimum Number of Taps to Open to Water a Garden.rs
impl Solution {
    pub fn min_taps(n: i32, ranges: Vec<i32>) -> i32 {
        let mut coverage = vec![0; (n + 1) as usize];

        // Update coverage array based on tap ranges
        for i in 0..ranges.len() {
            if ranges[i] == 0 {
                continue;
            }
            let left = std::cmp::max(0, i as i32 - ranges[i]) as usize;
            coverage[left] = std::cmp::max(coverage[left], (i as i32 + ranges[i]) as usize);
        }

        let (mut end, mut farthest_can_reach, mut tap_count) = (0, 0, 0);

        // Traverse the garden and calculate required taps
        for i in 0..=n as usize {
            if i > end {
                if farthest_can_reach <= end {
                    return -1;
                }
                end = farthest_can_reach;
                tap_count += 1;
            }
            farthest_can_reach = std::cmp::max(farthest_can_reach, coverage[i]);
        }

        // Return the total taps required, considering the endpoint
        tap_count + if end < n as usize { 1 } else { 0 }
    }
}
